readers/common/src/main/java/com/criteo/hadoop/garmadon/protobuf/ProtoConcatenator.java

```diff
@@ -0,0 +1,152 @@
+        return concatInner(messages, (ignored) -> new HashMap<>(), (entry, eventMap) -> {
+            eventMap.put(entry.getKey().getName(), entry.getValue());
+        });
+    }
+
+    /**
+     * Build a dynamic message builder based on a list of fields. Fields will be numbered in the order they were
+     * provided, starting from 1
+     *
+     * @param msgName   Name of the output message
+     * @param fields    Fields to be added to the output message definition
+     * @return          A builder able to fill a message made of all input fields
+     * @throws Descriptors.DescriptorValidationException    In case of a bug (shouldn't happen)
+     */
+    public static DynamicMessage.Builder buildMessageBuilder(String msgName, Collection<Descriptors.FieldDescriptor> fields)
+            throws Descriptors.DescriptorValidationException {
+        MessageDefinition.Builder msgDef = MessageDefinition.newBuilder(msgName);
+
+        int currentIndex = 1;
+
+        for (Descriptors.FieldDescriptor fieldDescriptor: fields) {
+            String label;
+
+            if (fieldDescriptor.isRepeated()) {
+                label = "repeated";
+            }
+            else
+                label = (fieldDescriptor.isRequired()) ? "required" : "optional";
+
+            msgDef.addField(label,
+                    fieldDescriptor.getType().toString().toLowerCase(), fieldDescriptor.getName(), currentIndex++);
+        }
+
+        DynamicSchema.Builder schemaBuilder = DynamicSchema.newBuilder();
+        schemaBuilder.addMessageDefinition(msgDef.build());
+
+        DynamicSchema schema = schemaBuilder.build();
+
+        return schema.newMessageBuilder(msgName);
+    }
+
+    /**
+     * Concatenate Protobuf messages and feeds them to parameterizable consumers
+     * /!\ Doesn't handle embedded objects and repeated fields /!\
+     *
+     * @param messages              Messages to be concatenated
+     * @param messageBuilder        Callback invoked with all input fields. Returns the object to be filled by
+     *                              contentsConsumer.
+     * @param contentsConsumer      Callback invoked for every output field + contents, along with the messageBuilder.
+     *                              Returns false if something wrong happened.
+     *
+     * @return                      Whether all operations completed successfully
+     */
+    private static <MessageType> MessageType concatInner(Collection<Message> messages,
+                   Function<Collection<Descriptors.FieldDescriptor>, MessageType> messageBuilder,
+                   BiConsumer<Map.Entry<Descriptors.FieldDescriptor, Object>, MessageType> contentsConsumer) {
+        Collection<Map.Entry<Descriptors.FieldDescriptor, Object>> allFields = new HashSet<>();
+
+        for (Message message : messages) {
+            allFields.addAll(message.getAllFields().entrySet());
+        }
+
+        Collection<Descriptors.FieldDescriptor> allKeys = new LinkedList<>();
```

Comment:
ArrayList?

Reply:
I don't really mind either way, not sure it makes a big difference given I'm only doing additions and traversals